### PR TITLE
Implement Sleeper Agent Poisoning Attack in TensorFlow

### PR DESCRIPTION
**Pull request by monshri**
# Description
Supporting Sleeper Agent Attack in TensorFlow framework

Fixes # (issue)

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 1]
```diff
@@ -200,7 +206,13 @@ def poison(  # type: ignore
 
         if self.verbose > 0:
             print("Best B-score:", best_B)
-        x_train[self.indices_target[best_indices_poison]] = np.transpose(best_x_poisoned, [0, 2, 3, 1])
+        #         pdb.set_trace()
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:209]:
```suggestion
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 2]
```diff
@@ -283,49 +316,146 @@ def create_model(
...
+        from art.estimators.classification.pytorch import PyTorchClassifier
+        from art.estimators.classification import TensorFlowV2Classifier
+
+        if isinstance(self.substitute_classifier, PyTorchClassifier):
+            import torch
+            from torch import nn
+            from torch.utils.data import TensorDataset, DataLoader
+            import torchvision
+
+            device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:328]: Should we use the same device as for the substitute classifier or ask for the GPU ID as input?

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 3]
```diff
@@ -200,7 +206,13 @@ def poison(  # type: ignore
 
         if self.verbose > 0:
             print("Best B-score:", best_B)
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:208]: Can we use logging instead of print?

Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:208]: Sure

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 4]
```diff
@@ -373,27 +502,46 @@ def select_poison_indices(
...
+            import tensorflow as tf
+
+            grad_norms = []
+            for i in range(len(x_samples) - 1):
+                image = tf.constant(x_samples[i : i + 1])
+                label = tf.constant(y_samples[i : i + 1])
+                with tf.GradientTape() as t:  # pylint: disable=C0103
+                    t.watch(classifier.model.weights)
+                    output = classifier.model(image, training=False)
+                    loss = classifier.model.compiled_loss(label, output)  # type: ignore
```
[the remark this answers is not in the document]
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:535]: We were getting issues in style check. So, it was suggested by you to use # type ignore in this specific case

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 5]
```diff
@@ -283,49 +316,146 @@ def create_model(
...
+                    optimizer.step()
+                    _, predicted = torch.max(outputs.data, 1)
+                    total += labels.size(0)
+                    accuracy += (predicted == labels).sum().item()
+                    running_loss += loss.item()
+                if (epoch % 5 == 0) or epoch == (epochs - 1):
+                    train_accuracy = 100 * accuracy / total
+                    print("Epoch {} train accuracy: {}".format(epoch, train_accuracy))  # pylint: disable=C0209
+            test_accuracy = self.test_accuracy(model, dataloader_test)
+            print("Final test accuracy: {}".format(test_accuracy))  # pylint: disable=C0209
```
[the remark this answers is not in the document]
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:368]: Sure

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 2 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:327]: I think we can keep the same device for the substitute classifier as otherwise it will involve switching between the GPUs?

Comment (beat-buesser) [reply art/attacks/poisoning/sleeper_agent_attack.py:327]: @monshri Yes, let's use the same device.

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/poisoning/sleeper_agent_attack.py:535]: Can you resolve it by giving the two variables `loss` in the two if-blocks for TensorFlow and PyTorch specific names like `loss_tf` and `loss_pt`?

[the remark this answers is not in the document]
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:535]: Sure

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 2 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:327]: Sure

File: requirements_test.txt [hunk 6]
```diff
@@ -19,6 +19,7 @@ opencv-python
 sortedcontainers==2.4.0
 h5py==3.7.0
 
+
```
Comment (beat-buesser) [inline requirements_test.txt:22]:
```suggestion
```

File: tests/attacks/poison/test_sleeper_agent_attack.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+import logging
+import numpy as np
+import pytest
+from PIL import Image
+
+from art.attacks.poisoning.sleeper_agent_attack import SleeperAgentAttack
+from art.utils import to_categorical
+from numpy import asarray
+
+from tests.utils import ARTTestException
```
Comment (beat-buesser) [inline tests/attacks/poison/test_sleeper_agent_attack.py:29]: Please sort imports according to PEP8.

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 8]
```diff
@@ -136,21 +142,28 @@ def poison(  # type: ignore
         :return: x_train, y_train and indices of poisoned samples.
                  Here, x_train are the samples selected from target class
                  in training data.
-        """
+        """
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:145]:
```suggestion
        """
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 9]
```diff
@@ -283,117 +328,169 @@ def create_model(
...
+            predictions = model_new.predict(x_test)
+            accuracy = np.sum(np.argmax(predictions, axis=1) == np.argmax(y_test, axis=1)) / len(y_test)
+            print("Accuracy of substitute model after retraining: {}%".format(accuracy * 100))
+        
+        elif isinstance(self.substitute_classifier, TensorFlowV2Classifier):
+            import tensorflow as tf
+            from tqdm.keras import TqdmCallback
+
+            # Tweaked the model from https://github.com/calmisential/TensorFlow2.0_ResNet
+            # MIT License
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:354]: Please include their license text at the top of this file below the license of ART.

File: tests/attacks/poison/test_sleeper_agent_attack.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+logger = logging.getLogger(__name__)
+
+
+
+@pytest.mark.only_with_platform("pytorch", "tensorflow2")
+def test_poison(art_warning, get_default_mnist_subset, image_dl_estimator,framework):
+    try:
+        (x_train, y_train), (x_test, y_test) = get_default_mnist_subset
+        if framework == 'pytorch':
+            x_train = np.transpose(x_train, (0, 2, 3, 1))
```
Comment (beat-buesser) [inline tests/attacks/poison/test_sleeper_agent_attack.py:40]: Can we test PyTorch with channels first images?

File: tests/attacks/poison/test_sleeper_agent_attack.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        classifier, _ = image_dl_estimator()
+        x_train, y_train = x_train[:1000], y_train[:1000]
+        max_ = 1
+        min_ = 0
+        mean = np.mean(x_train, axis=(0, 1, 2, 3))
+        std = np.std(x_train, axis=(0, 1, 2, 3))
+        x_train = (x_train - mean) / (std + 1e-7)
+        x_test = (x_test - mean) / (std + 1e-7)
+        min_ = (min_ - mean) / (std + 1e-7)
+        max_ = (max_ - mean) / (std + 1e-7)
```
Comment (beat-buesser) [inline tests/attacks/poison/test_sleeper_agent_attack.py:50]: Can we test without pre-normalising?

File: tests/attacks/poison/test_sleeper_agent_attack.py (hunk 10 above)
Comment (monshri) [reply tests/attacks/poison/test_sleeper_agent_attack.py]: Added Functionality provided for handling both channel first and channel last

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 9 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py]: Removing this license as using art architecture for training now.

File: tests/attacks/poison/test_sleeper_agent_attack.py (hunk 7 above)
Comment (monshri) [reply tests/attacks/poison/test_sleeper_agent_attack.py:29]: Done

**Review by beat-buesser — APPROVED** [review]

**Comment by beat-buesser** [comment]
Hi @monshri Thank you very much for your contribution!
